loki/transformations/temporaries/stack_allocator.py

```diff
@@ -0,0 +1,1372 @@
+            The driver subroutine to get the stack_variables
+        stack_dict : dict
+            dict that maps dtype and kind to an expression for the required stack size
+        successors : list of :any:`Item`
+            The items corresponding to successor routines called from :data:`routine`
+        """
+
+        # block variables
+        kgpblock = self._get_int_var(name=self.block_dim.size, scope=routine)
+        jgpblock = self._get_int_var(name=self.block_dim.index, scope=routine)
+
+        stack_vars = []
+        stack_arg_dict = {}
+        assignments = []
+        deallocs = []
+        pragma_string = ''
+        pragma_data_start = None
+        for dtype in stack_dict:
+            for kind in stack_dict[dtype]:
+                # start integer names in the driver with 'J'
+                stack_size_name = self._get_stack_int_name('J', dtype, kind, 'STACK_SIZE')
+                stack_size_var = self._get_int_var(name=stack_size_name, scope=routine)
+
+                stack_used_name = self._get_stack_int_name('J', dtype, kind, 'STACK_USED')
+                stack_used_var = self._get_int_var(name=stack_used_name, scope=routine)
+
+                # create the stack variable and its type with the correct shape
+                stack_var = self._get_stack_var(routine, dtype, kind)
+
+                stack_type = stack_var.type.clone(shape=(RangeIndex((None,None)), RangeIndex((None,None))),
+                                                  allocatable=True)
+                stack_var = stack_var.clone(type=stack_type)
+
+                stack_alloc = Allocation(variables=(stack_var.clone(dimensions=(stack_dict[dtype][kind], kgpblock)),))
+                stack_dealloc = Deallocation(variables=(stack_var.clone(dimensions=None),))
+
+                # add the variables to the stack_arg_dict with dimensions (:,j_block)
+                stack_arg_dict.setdefault(dtype, {})
+                stack_arg_dict[dtype][kind] = (stack_size_var,
+                                               stack_var.clone(dimensions=(RangeIndex((None,None)), jgpblock,)),
+                                               stack_used_var)
+                stack_var = stack_var.clone(dimensions=stack_type.shape)
+
+                stack_used_var_init = Assignment(lhs=stack_used_var, rhs=IntLiteral(1))
+                # create stack_vars pair and assignment of the size variable
+                stack_vars += [stack_size_var, stack_var, stack_used_var]
+                assignments += [Assignment(lhs=stack_size_var,
+                                           rhs=stack_dict[dtype][kind]), stack_alloc, stack_used_var_init]
+                deallocs += [stack_dealloc]
+                pragma_string += f'{stack_var.name}, '
+
+        # add to routine
+        routine.variables = routine.variables + as_tuple(stack_vars)
+        nodes_to_add = assignments
+
+        if pragma_string:
+            pragma_string = pragma_string[:-2].lower()
```

Comment:
[optional] You could avoid this if you collected variable names in a list and then just did `pragma_string = ', '.join(stack_data_vars)`